### PR TITLE
 Add caching and wildcard support to ground items

### PR DESCRIPTION
**Pull request by deathbeam**
- Use caching for getting the data of each ground item
- Use collectors for reducing items to sum prices and quantities
- Add wilcard matching support
- Use cache for wildcard matching results

Fixes: #658

Signed-off-by: Tomas Slusny <slusnucky@gmail.com>

Preview of the wildcards:
screenshot 2018-02-27 21 56 51
screenshot 2018-02-27 21 56 28

### CODE REVIEW

**Comment by deathbeam** [comment]
Updated OP with screenshots of the wildcards (i have GE prices disabled to fit on screenie, but they work)

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemLoader.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	{
+		return executorService.submit(() -> loadGroundItem(item));
+	}
+
+	private GroundItem loadGroundItem(@Nonnull final GroundItemCacheKey item)
+	{
+
+		final int itemId = item.getId();
+		final int itemQuantity = item.getQuantity();
+		final ItemComposition itemDefinition = itemManager.getItemComposition(itemId);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemLoader.java]: It is not safe to call this from threads

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 2]
```diff
@@ -36,22 +48,100 @@
...
+			hiddenItemAssociations.cleanUp();
+		}
+
+		if (highlightedItemAssociations != null)
+		{
+			highlightedItemAssociations.cleanUp();
+		}
+
+		// create new association caches
+		hiddenItemAssociations = CacheBuilder.newBuilder()
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Create these in startUp and then invalidate them here

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemLoader.java (hunk 1 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemLoader.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java (hunk 2 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Can't do, they are created from parsed highlighted and hidden item lists.

**Pull request closed without merging**